core/auth_provider_okta.go

```diff
@@ -0,0 +1,311 @@
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(b, p)
+	if err != nil {
+		return err
+	}
+
+	if p.ClientID == "" {
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `client_id' value")
+	}
+
+	if p.ClientSecret == "" {
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `client_secret' value")
+	}
+
+	if p.OktaDomain == "" {
+		// ex: p.OktaDomain = "https://dev-xyz.okta.com/" 
+		// p.OktaIssuer = "https://dev-xyz.okta.com/oauth2/default" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `okta_domain' value")
+	}
+
+	if p.DeploymentURI == "" {
+		// ex: p.DeploymentURI = "https://shield.starkandwayne.com" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `deployment_uri' value")
+	}
+
+	if p.AuthorizationServer == "" {
+		// ex: p.AuthorizationServer = "default" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `authorization_server' value")
+	}
+
+
+	p.OktaDomain = strings.TrimSuffix(p.OktaDomain, "/")
+	p.DeploymentURI = strings.TrimSuffix(p.DeploymentURI, "/")
+
+
+	p.properties = util.StringifyKeys(raw).(map[string]interface{})
+
+	return nil
+}
+
+func (p *OktaAuthProvider) WireUpTo(c *Core) {
+	p.core = c
+}
+
+func (p *OktaAuthProvider) ReferencedTenants() []string {
+	ll := make([]string, 0)
+	for _, m := range p.Mapping {
+		ll = append(ll, m.Tenant)
+	}
+	return ll
+}
+
+func (p *OktaAuthProvider) Initiate(r *route.Request) {
+
+	var state = randstr.Hex(16)
+
+	var redirectPath string
+
+	q := r.Req.URL.Query()
+	q.Add("client_id", p.ClientID)
+	q.Add("response_type", "code")
+	q.Add("response_mode", "query")
+	q.Add("scope", "openid profile email groups")
+	q.Add("redirect_uri", p.DeploymentURI + "/auth/okta/redir") //ex: https://shield.starkandwayne.com/auth/okta/redir
+	q.Add("state", state)
+	// q.Add("nonce", nonce)
+
+	redirectPath = p.OktaDomain + "/oauth2/" + p.AuthorizationServer + "/v1/authorize?" + q.Encode()
+	//ex: "https://dev-xyz.okta.com/oauth2/default/v1/authorize?" 
+
+	r.Redirect(302, redirectPath)
+}
+
+func (p *OktaAuthProvider) HandleRedirect(r *route.Request) *db.User {
+	
+	// The login is achieved through the authorization code flow, where the user 
+	// is redirected to the Okta-Hosted login page. After the user authenticates 
+	// they are redirected back to the application with an access code that is 
+	// then exchanged for an access_token.
+	
+	// log.Debugf("DEBUG::r *route.Request is [GET /auth/okta/redir]")
+
+	q := r.Req.URL.Query()
+	q.Add("grant_type", "authorization_code")
+	q.Set("code", r.Param("code", ""))
+	q.Add("redirect_uri", p.DeploymentURI + "/auth/okta/redir") //ex: https://shield.starkandwayne.com/auth/okta/redir
+	q.Add("response_type", "token")
+
+	url := p.OktaDomain + "/oauth2/" + p.AuthorizationServer + "/v1/token?" + q.Encode()
+
+	authHeader := base64.StdEncoding.EncodeToString([]byte(p.ClientID + ":" + p.ClientSecret))
```

Comment:
Use https://pkg.go.dev/net/http#Request.SetBasicAuth to set the `Authorization` header instead.

Reply:
SetBasicAuth uses `r.Header.Set("Authorization", "Basic "+basicAuth(username, password))` but we have `base64.StdEncoding.EncodeToString([]byte(p.ClientID + ":" + p.ClientSecret))`

Reply:
Then we should keep it as it is, nice catch.